src/index.js

```diff
@@ -0,0 +1,253 @@
+   */
+  start() {
+    assert(!this.running, 'Already running');
+    this.debug('starting');
+    this.running = true;
+    this.recycle();
+  }
+
+  async stop() {
+    assert(this.running, 'Not running');
+    this.debug('stopping');
+    this.running = false;
+    this.recycle();
+
+    // wait until all existing connections are finished
+    const unfinished = this.connections.filter(conn => conn.state !== 'finished');
+    if (unfinished.length > 0) {
+      await Promise.all(unfinished.map(
+        conn => new Promise(resolve => { conn.once('finished', resolve); })));
+    }
+  }
+
+  /**
+   * Create a new connection, retiring any existing connection.
+   */
+  recycle() {
+    this.debug('recycling');
+
+    if (this.connections.length) {
+      const currentConn = this.connections[0];
+      currentConn.retire();
+    }
+
+    if (this.running) {
+      const newConn = new PulseConnection(this, ++this.connectionCounter);
+      newConn.once('connected', () => {
+        this.emit('newConnection', newConn);
+      });
+      newConn.once('finished', () => {
+        this.connections = this.connections.filter(conn => conn.id !== newConn.id);
+      });
+      this.connections.unshift(newConn);
+    }
+  }
+}
+
+exports.PulseConnector = PulseConnector;
+
+/**
+ * A single connection to a pulse server.  This is a thin wrapper around a raw
+ * AMQP connection, instrumented to inform the parent PulseConnector of failures
+ * and trigger a reconnection.  It is possible to have multiple PulseConnection
+ * objects in the same process at the same time, while one is being "retired" but
+ * is lingering around to send ack's for any in-flight message handlers.
+ *
+ * The instance's `amqp` property is the amqp connection object.  In the event of any
+ * issues with the connection, call the instance's `failed` method.  This will initiate
+ * a retirement of the connection and creation of a new connection.
+ *
+ * The instance will emit a `connected` event when it connects to the pulse server.
+ * This event occurs before the connection is provided to a user, so it is only
+ * of interest to the PulseConnector class.
+ *
+ * This instance will emit a `retiring` event just before it is retired.  Users
+ * should cancel consuming from any channels, as a new connection will soon
+ * begin consuming.  Errors from such cancellations should be logged and
+ * ignored.  This connection will remain open for 30 seconds to allow any
+ * in-flight message processing to complete.
+ *
+ * The instance will emit `finished` when the connection is finally closed.
+ *
+ * A connection's state can be one of
+ *
+ *  - connecting -- waiting for a connection to complete
+ *  - connected -- connection is up and running
+ *  - retiring -- in the process of retiring
+ *  - finished -- no longer connected
+ *
+ *  Note that an instance that fails to connect will skip from `connecting` to
+ *  `retiring`.
+ *
+ */
+class PulseConnection extends events.EventEmitter {
```

Comment:
This needs a method for closing and not reconnecting...

IMO, you could do away with the connector logic and put it all in here. Disconnection should be detected by:
 1. the amqp connection error event
 2. the amqp connection closed event (without a call to `PulseConnection.close()` which would make this okay)